src/tdyio.c

```diff
@@ -11,43 +12,107 @@ PetscErrorCode TDyIOCreate(TDyIO *_io) {
+
+  PetscInt dim,istart,iend,numCell,numVert,numCorner;
+
+  zonalVarNames[0] = tdy->io->zonalVarNames[0];
+
+  if (tdy->io->format == ExodusFormat) {
+    strcpy(tdy->io->filename, "out.exo");
+    char *ofilename = tdy->io->filename;
+    ierr = TdyIOInitializeExodus(ofilename,zonalVarNames,dm,num_vars);CHKERRQ(ierr);
+  }
+  else if (tdy->io->format == HDF5Format) {
+    strcpy(tdy->io->filename, "out.h5");
+    char *ofilename = tdy->io->filename;
+    numCorner = TDyGetNumberOfCellVertices(dm);
+    ierr = DMGetDimension(dm,&dim);CHKERRQ(ierr); 
+    ierr = DMPlexGetHeightStratum(dm,3,&istart,&iend);CHKERRQ(ierr); //just for box example?
+    numVert = iend-istart;
+    ierr = DMPlexGetHeightStratum(dm,0,&istart,&iend);CHKERRQ(ierr);
+    numCell = iend-istart;
+    
+      
+    ierr = TdyIOInitializeHDF5(ofilename,dm);CHKERRQ(ierr);
+    ierr = TDyIOWriteXMFHeader(numCell,dim,numVert,numCorner);CHKERRQ(ierr);
+  }
+    
   PetscFunctionReturn(0);
 }
 
 PetscErrorCode TDyIOWriteVec(TDy tdy){
   PetscErrorCode ierr;
-  char *zonalVarNames[1];
- 
-  char *ofilename = tdy->io->exodus_filename;
+  PetscInt numCell,istart,iend;
+  
   int num_vars = tdy->io->num_vars;
   Vec v = tdy->solution;
   DM dm = tdy->dm;
   PetscReal time = tdy->ti->time;
-  zonalVarNames[0] = tdy->io->zonalVarNames[0];
-  
+ 
   if (tdy->io->format == PetscViewerASCIIFormat) {
-    TDyIOPrintVec(v, time);
+    ierr = TDyIOWriteAsciiViewer(v, time);CHKERRQ(ierr);
   }
-  if (tdy->io->format == ExodusFormat){
-    if (tdy->io->num_times == 0) {
-      TdyIOInitializeExodus(ofilename,zonalVarNames,dm,num_vars);
-      ierr = PetscObjectSetName((PetscObject) v,  "Soln");CHKERRQ(ierr);
-    }
-    TdyIOAddExodusTime(ofilename,time,tdy->io);
-    TdyIOWriteExodusVar(ofilename,v,tdy->io);
+  else if (tdy->io->format == ExodusFormat) {
+    char *ofilename = tdy->io->filename;
+
+    ierr = TdyIOAddExodusTime(ofilename,time,tdy->io);CHKERRQ(ierr);
+    ierr = TdyIOWriteExodusVar(ofilename,v,tdy->io);CHKERRQ(ierr);
+  }
+  else if (tdy->io->format == HDF5Format) {
+    char *ofilename = tdy->io->filename;
+    ierr = DMPlexGetHeightStratum(dm,0,&istart,&iend);CHKERRQ(ierr);
+    numCell = iend-istart;
+    ierr = TdyIOWriteHDF5Var(ofilename,v,time,numCell);CHKERRQ(ierr);
   }
+  else{
+    SETERRQ(PETSC_COMM_WORLD,PETSC_ERR_USER,"Unrecognized IO format, must call TDyIOSetMode");
+  }
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode TdyIOInitializeHDF5(char *ofilename, DM dm){
+  PetscViewer viewer; 
+  PetscErrorCode ierr;
+  PetscViewerFormat format;
+  format = PETSC_VIEWER_HDF5_XDMF;
+  
+  ierr = PetscViewerHDF5Open(PETSC_COMM_WORLD,ofilename,FILE_MODE_WRITE,&viewer);CHKERRQ(ierr);
+  ierr = PetscViewerPushFormat(viewer, format);CHKERRQ(ierr);
+  ierr = DMView(dm,viewer);CHKERRQ(ierr);
+  ierr = PetscViewerDestroy(&viewer);CHKERRQ(ierr);
+  
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode TdyIOWriteHDF5Var(char *ofilename, Vec U,PetscReal time,PetscInt numCell){   
```

Comment:
Instead of passing in `numCell`, you can compute ncells by `VecGetSize(U, &numCell)`.